src/esperanza/finalizationstate.cpp

```diff
@@ -629,21 +629,8 @@ Result FinalizationState::CalculateWithdrawAmount(const uint160 &validatorAddres
                 validatorAddress.GetHex());
   }
 
-  const auto &validator = it->second;
-
-  uint32_t endDynasty = validator.m_end_dynasty;
-
-  if (m_current_dynasty <= endDynasty) {
-    return fail(Result::WITHDRAW_BEFORE_END_DYNASTY,
-                /*log_errors=*/true,
-                "%s: Too early to withdraw, minimum expected dynasty for "
-                "withdraw is %d.\n",
-                __func__, endDynasty);
-  }
-
-  uint32_t endEpoch = m_dynasty_start_epoch.find(endDynasty + 1)->second;
-  uint32_t withdrawalEpoch = endEpoch + m_settings.withdrawal_epoch_delay;
-
+  const Validator &validator = it->second;
+  const uint32_t withdrawalEpoch = CalculateWithdrawEpoch(validator);
```

Comment:
According to proposed implementation, when `m_current_dynasty <= finalizer.m_end_dynasty`, this function will print "minimum expected epoch for withdraw is 4294967295.". it's incorrect, printed value must be `endEpoch + m_settings.withdrawal_epoch_delay`

Reply:
I didn't understand your comment, probably it's because we don't have `endEpoch` variable up to this point. 
If we have the following state:
```c++
m_current_epoch=3
m_current_dynasty=1
m_end_dynasty=3
withdrawal_epoch_delay=2
```
What's the value you say is a correct one to write in logs?

Reply:
Yeah sorry I was confused by all of these code movements again. I think proposed logging is exactly the same as it was.

But, what I'm worrying about, and why I'd been returning here again and again, is this 4294967295 number in the log. We do not really need to see epoch of dynasty which is in the future, right? It's always the 4294967295 because we have no chance to predict it. Instead, I'd probably implement logging in this way:

```C++
  if (m_current_epoch < withdrawalEpoch) {
    if (withdrawalEpoch == MAX_END_DYNASTY) {
      return fail(Result::WITHDRAW_TOO_EARLY,
                  /*log_errors=*/true,
                  "%s: Too early to withdraw, withdraw opens on dynasty %d"
                  __func__, validator.m_end_dynasty + 1);

    } else {
      return fail(Result::WITHDRAW_TOO_EARLY,
                  /*log_errors=*/true,
                  "%s: Too early to withdraw, minimum expected epoch for "
                  "withdraw is %d.\n",
                  __func__, withdrawalEpoch);
    }
  }
```

While been writing this code, I also find that function CalculateWithdraw**Epoch** returns MAX_END_**DYNASTY**. It's nit but would be fine to not pass confusing names to the master, what to you think?

Reply:
If you use our client, it would be very hard to reproduce this log.

At first, RPC won't allow you to create withdraw earlier than you should. Then when the node broadcasts the tx to other peers they should be 1 epoch behind you to end up in this scenario (but it can be only 1 block if you created the withdraw on the first block). Then even if this happens, it will "might" confuse (because I think it's OK what it shows now) a very limited number of people. 

About frequency. Once you decide to withdraw, at first you need to logout. The logout delay + withdraw delay would be quite high. I tend to think about several weeks (but it's TBD).

Because I think the current log is correct and a chance to see it very low I wouldn't extend this code because it's smth that we would need to support.